database2/ffboltdb/db.go

```diff
@@ -0,0 +1,1510 @@
+type bulkFetchDataSorter []bulkFetchData
+
+// Len returns the number of items in the slice.  It is part of the
+// sort.Interface implementation.
+func (s bulkFetchDataSorter) Len() int {
+	return len(s)
+}
+
+// Swap swaps the items at the passed indices.  It is part of the
+// sort.Interface implementation.
+func (s bulkFetchDataSorter) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+// Less returns whether the item with index i should sort before the item with
+// index j.  It is part of the sort.Interface implementation.
+func (s bulkFetchDataSorter) Less(i, j int) bool {
+	if s[i].blockFileNum < s[j].blockFileNum {
+		return true
+	}
+	if s[i].blockFileNum > s[j].blockFileNum {
+		return false
+	}
+
+	return s[i].fileOffset < s[j].fileOffset
+}
+
+// makeDbErr creates a database.Error given a set of arguments.
+func makeDbErr(c database.ErrorCode, desc string, err error) database.Error {
+	return database.Error{ErrorCode: c, Description: desc, Err: err}
+}
+
+// convertErr converts the passed bolt error into a database error with an
+// equivalent error code  and the passed description.  It also sets the passed
+// error as the underlying error.
+func convertErr(desc string, boltErr error) database.Error {
+	// Use the driver-specific error code by default.  The code below will
+	// update this with the converted bolt error if it's recognized.
+	var code = database.ErrDriverSpecific
+
+	switch boltErr {
+	// Database open/create errors.
+	case bolt.ErrDatabaseNotOpen:
+		code = database.ErrDbNotOpen
+	case bolt.ErrInvalid:
+		code = database.ErrInvalid
+
+	// Transaction errors.
+	case bolt.ErrTxNotWritable:
+		code = database.ErrTxNotWritable
+	case bolt.ErrTxClosed:
+		code = database.ErrTxClosed
+
+	// Value/bucket errors.
+	case bolt.ErrBucketNotFound:
+		code = database.ErrBucketNotFound
+	case bolt.ErrBucketExists:
+		code = database.ErrBucketExists
+	case bolt.ErrBucketNameRequired:
+		code = database.ErrBucketNameRequired
+	case bolt.ErrKeyRequired:
+		code = database.ErrKeyRequired
+	case bolt.ErrKeyTooLarge:
+		code = database.ErrKeyTooLarge
+	case bolt.ErrValueTooLarge:
+		code = database.ErrValueTooLarge
+	case bolt.ErrIncompatibleValue:
+		code = database.ErrIncompatibleValue
+	}
+
+	return database.Error{ErrorCode: code, Description: desc, Err: boltErr}
+}
+
+// cursor is an internal type used to represent a cursor over key/value pairs
+// and nested buckets of a bucket and implements the database.Cursor interface.
+type cursor struct {
+	bucket     *bucket
+	boltCursor *bolt.Cursor
+}
+
+// Enforce cursor implements the database.Cursor interface.
+var _ database.Cursor = (*cursor)(nil)
+
+// Bucket returns the bucket the cursor was created for.
+//
+// This function is part of the database.Cursor interface implementation.
+func (c *cursor) Bucket() database.Bucket {
+	// Ensure transaction state is valid.
+	if err := c.bucket.tx.checkClosed(); err != nil {
+		return nil
```

Comment:
Well callers should never be closing the db and/or transaction out from under themselves, but it has the same semantics as bolt - it'll panic.